module/bdev/raid/bdev_raid.c

```diff
@@ -2196,6 +2271,384 @@ _raid_bdev_remove_base_bdev(struct raid_base_bdev_info *base_info,
+	if (!base_info->delta_bitmap) {
+		SPDK_ERRLOG("Unable to allocate delta bitmap for base bdev %s\n", base_bdev_name);
+		return -ENOMEM;
+	}
+
+	base_info->state = BASE_BDEV_STATE_FAULTY_STOPPED;
+
+	ctx->base_info = base_info;
+	ctx->cb_fn = cb_fn;
+	ctx->cb_arg = cb_arg;
+
+	return spdk_bdev_quiesce(bdev, &g_raid_if, _raid_bdev_base_bdev_stop_faulty_state, ctx);
+
+
+	return 0;
+}
+
+/*
+ * brief:
+ * raid_bdev_get_base_bdev_delta_bitmap function is the method to get the delta map of a base bdev that is
+ * in faulty state.
+ * If the base bdev is not part of any raid, or if the base bdev is not in faulty state, NULL is returned.
+ * params:
+ * base_bdev_name - name of the base bdev
+ * returns:
+ * delta map - success
+ * NULL - failure
+ */
+struct spdk_bit_array *raid_bdev_get_base_bdev_delta_bitmap(char *base_bdev_name)
+{
+	struct raid_base_bdev_info *base_info;
+
+	assert(spdk_get_thread() == spdk_thread_get_app_thread());
+
+	base_info = raid_bdev_find_faulty_base_info_by_name(base_bdev_name);
+
+	if (!base_info) {
+		SPDK_ERRLOG("bdev '%s' not a faulty base info of any RAID\n", base_bdev_name);
+		return NULL;
+	}
+
+	if (base_info->state != BASE_BDEV_STATE_FAULTY_STOPPED) {
+		SPDK_ERRLOG("Delta bitmap of base bdev '%s' is not stopped\n", base_bdev_name);
+		return NULL;
+	}
+
+	return base_info->delta_bitmap;
+}
+
+/*
+ * brief:
+ * raid_bdev_clear_base_bdev_faulty_state function is the method to clear the faulty state of a
+ * base bdev and so to delete the delta bitmap.
+ * If the base bdev is not part of any raid, or if the base bdev is not in faulty state, false is
+ * returned.
+ * params:
+ * base_bdev_name - name of the base bdev
+ * returns:
+ * true - success
+ * false - failure
+ */
+int
+raid_bdev_clear_base_bdev_faulty_state(char *base_bdev_name, raid_bdev_destruct_cb cb_fn,
+				       void *cb_arg)
+{
+	struct raid_base_bdev_info *base_info;
+	int rc;
+
+	assert(spdk_get_thread() == spdk_thread_get_app_thread());
+
+	base_info = raid_bdev_find_faulty_base_info_by_name(base_bdev_name);
+
+	if (!base_info) {
+		SPDK_ERRLOG("bdev '%s' not a faulty base info of any RAID\n", base_bdev_name);
+		return -ENODEV;
+	}
+
+	if (base_info->state == BASE_BDEV_STATE_NONE) {
+		SPDK_ERRLOG("Base bdev %s not in faulty state", base_bdev_name);
+		return -EPERM;
+	}
+
+	rc = _raid_bdev_base_bdev_clear_faulty_state(base_info, cb_fn, cb_arg);
+
+	return rc;
+}
+
+uint64_t
+raid_bdev_region_size_base_bdev_delta_bitmap(char *base_bdev_name)
+{
+	struct raid_base_bdev_info *base_info;
+	struct spdk_bdev *bdev;
+
+	assert(spdk_get_thread() == spdk_thread_get_app_thread());
+
+	base_info = raid_bdev_find_faulty_base_info_by_name(base_bdev_name);
+
+	if (!base_info) {
+		SPDK_ERRLOG("bdev '%s' not a faulty base info of any RAID\n", base_bdev_name);
+		return 0;
+	}
+
+	bdev = &base_info->raid_bdev->bdev;
+	return bdev->optimal_io_boundary * bdev->blocklen;
+}
```

Comment:
I just realize the region size is not cluster size. Can you elaborate a bit on `optimal I/O boundary` here?

Reply:
`optimal_io_boundary` is a concept of all bdevs, cluster size only for logical volumes.
However, for logical volumes, bdev's `optimal_io_boundary` is cluster size:
https://github.com/longhorn/spdk/blob/a6478cde7e0cff2fb09992868308a7387aa5202a/module/bdev/lvol/vbdev_lvol.c#L1150

Reply:
I see. It is clear. Thank you.